File: Introduction to Robotics Course/Chapter3/Lesson1/Chapter3_Lesson1.java
abstract class Robot {
    String name;
    int dofBase, dofJoints, nAgents;
    boolean humanoid;

    Robot(String name, int dofBase, int dofJoints, int nAgents, boolean humanoid) {
        this.name = name; this.dofBase = dofBase; this.dofJoints = dofJoints;
        this.nAgents = nAgents; this.humanoid = humanoid;
    }

    int dofTotal() { return dofBase + dofJoints; }

    String family() {
        if (nAgents > 1) return "Swarm";
        if (humanoid) return "Humanoid";
        if (dofBase > 0 && dofJoints == 0) return "Mobile robot";
        if (dofBase == 0 && dofJoints > 0) return "Manipulator";
        if (dofBase > 0 && dofJoints > 0) return "Mobile manipulator";
        return "Other";
    }
}

public class TaxonomyDemo {
    public static void main(String[] args) {
        Robot arm = new Robot("6R arm", 0, 6, 1, false){};
        Robot mobile = new Robot("Wheeled base", 3, 0, 1, false){};
        Robot humanoid = new Robot("Humanoid", 6, 22, 1, true){};
        Robot swarm = new Robot("Swarm", 3, 0, 50, false){};

        Robot[] robots = {arm, mobile, humanoid, swarm};
        for (Robot r : robots) {
            System.out.println(r.name + " => " + r.family() + " (DoF " + r.dofTotal() + ")");
        }
    }
}
      

File: Introduction_to_Robotics_Course/Chapter3/Lesson1/Chapter3_Lesson1.py
from dataclasses import dataclass
from typing import List, Optional

@dataclass
class Robot:
    name: str
    dof_base: int
    dof_joints: int
    n_agents: int = 1  # >1 indicates swarm
    humanoid: bool = False

    @property
    def dof_total(self) -> int:
        return self.dof_base + self.dof_joints

    def family(self) -> str:
        if self.n_agents > 1:
            return "Swarm"
        if self.humanoid:
            return "Humanoid"
        if self.dof_base > 0 and self.dof_joints == 0:
            return "Mobile robot"
        if self.dof_base == 0 and self.dof_joints > 0:
            return "Manipulator"
        if self.dof_base > 0 and self.dof_joints > 0:
            return "Mobile manipulator"
        return "Other"

robots = [
    Robot("6R arm", dof_base=0, dof_joints=6),
    Robot("Differential drive base", dof_base=3, dof_joints=0),
    Robot("Humanoid example", dof_base=6, dof_joints=22, humanoid=True),
    Robot("Micro-swarm", dof_base=3, dof_joints=0, n_agents=50)
]

for r in robots:
    print(r.name, "=>", r.family(), ", DoF total:", r.dof_total)
      